File: MyProject/Source/MyProject/Private/Game/Libs/FrameHandle/FrameTimerMgr.cpp
#include "MyProject.h"
#include "FrameTimerMgr.h"

FrameTimerMgr::FrameTimerMgr()
{

}

void FrameTimerMgr::addObject(IDelayHandleItem delayObject, float priority = 0.0f)
{
	// 检查当前是否已经在队列中
	if (m_timerLists.IndexOf(delayObject as FrameTimerItem) == -1)
	{
		if (bInDepth())
		{
			base.addObject(delayObject, priority);
		}
		else
		{
			m_timerLists.Add(delayObject as FrameTimerItem);
		}
	}
}

void FrameTimerMgr::delObject(IDelayHandleItem delayObject)
{
	// 检查当前是否在队列中
	if (m_timerLists.IndexOf(delayObject as FrameTimerItem) != -1)
	{
		(delayObject as FrameTimerItem).m_disposed = true;
		if (bInDepth())
		{
			base.addObject(delayObject);
		}
		else
		{
			foreach(FrameTimerItem item in m_timerLists)
			{
				if (UtilApi.isAddressEqual(item, delayObject))
				{
					m_timerLists.Remove(item);
					break;
				}
			}
		}
	}
}

void FrameTimerMgr::Advance(float delta)
{
	incDepth();

	foreach(FrameTimerItem timerItem in m_timerLists)
	{
		if (!timerItem.getClientDispose())
		{
			timerItem.OnFrameTimer();
		}
		if (timerItem.m_disposed)
		{
			delObject(timerItem);
		}
	}

	decDepth();
}

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/FrameHandle/FrameTimerMgr.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DelayHandle.DelayPriorityHandleMgrBase");

local M = GlobalNS.Class(GlobalNS.DelayPriorityHandleMgrBase);
M.clsName = "FrameTimerMgr";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mTimerList = GlobalNS.new(GlobalNS.MList);
    self.mDelList = GlobalNS.new(GlobalNS.MList);
end

function M:addObject(delayObject, priority)
    -- 检查当前是否已经在队列中
    if(not self.mTimerList:contains(delayObject)) then
        if(self:isInDepth()) then
            M.super.addObject(self, delayObject, priority);
        else
            self.mTimerList:add(delayObject);
        end
    end
end

function M:removeObject(delayObject)
    -- 检查当前是否在队列中
    if(self.mTimerList:contains(delayObject)) then
        delayObject.mIsDisposed = true;
        if(self:isInDepth()) then
            M.super.addObject(self, delayObject);
        else
            for key, item in ipairs(self.mTimerList.list()) do
                if item == delayObject then
                    self.mTimerList:remove(item);
                    break;
                end
            end
        end
    end
end

function M:Advance(delta)
    self.incDepth();

    for key, timerItem in ipairs(self.mTimerList.list()) do
        if not timerItem:getClientDispose() then
            timerItem:OnFrameTimer();
        end
        if timerItem.mIsDisposed then
            self:removeObject(timerItem);
        end
    end

    self:decDepth();
end

return M;